sklearn/metrics/cluster/supervised.py

```diff
@@ -188,33 +208,54 @@ def adjusted_rand_score(labels_true, labels_pred, max_n_classes=5000):
     adjusted_mutual_info_score: Adjusted Mutual Information
 
     """
-    labels_true, labels_pred = check_clusterings(labels_true, labels_pred)
-    n_samples = labels_true.shape[0]
-    classes = np.unique(labels_true)
-    clusters = np.unique(labels_pred)
+    if contingency is None:
+        labels_true, labels_pred = check_clusterings(labels_true, labels_pred)
+        n_samples = labels_true.shape[0]
+        n_classes = np.unique(labels_true).shape[0]
+        n_clusters = np.unique(labels_pred).shape[0]
+    elif isinstance(contingency, _data_matrix):
+        n_samples = contingency.nnz
+        n_classes, n_clusters = contingency.shape
+    else:
+        raise ValueError("'contingency' must be a sparse matrix or None")
+
     # Special limit cases: no clustering since the data is not split;
     # or trivial clustering where each document is assigned a unique cluster.
     # These are perfect matches hence return 1.0.
-    if (classes.shape[0] == clusters.shape[0] == 1 or
-            classes.shape[0] == clusters.shape[0] == 0 or
-            classes.shape[0] == clusters.shape[0] == len(labels_true)):
+    if (n_classes == n_clusters == 1 or
+                    n_classes == n_clusters == 0 or
+                    n_classes == n_clusters == n_samples):
         return 1.0
 
-    contingency = contingency_matrix(labels_true, labels_pred,
-                                     max_n_classes=max_n_classes)
+    # Compute contingency matrix if we weren't given it
+    if contingency is None:
+        contingency = contingency_matrix(labels_true, labels_pred,
+                                         max_n_classes=max_n_classes)
```

Comment:
This still builds dense contengency matrix by default. I would we should have a internal heuristic. If n_classes is small we use a dense array (because it's faster to compute) and if it's larger than than some threshold we switch to sparse.

Or we could decide to always build sparse contingency matrices internally to keep the code simpler and expect that a difference in speed will never exceed a couple of seconds for small to medium inputs while very large input will only be addressable via sparse contingency matrix anyway.